services/types.go

```diff
@@ -454,3 +459,77 @@ type HeartbeatService interface {
 	// Watch watches the heartbeats for a service
 	Watch() (xwatch.Watch, error)
 }
+
+// ElectionOptions configure specific election-scoped options.
+type ElectionOptions interface {
+	// Duration after which a call to Leader() will timeout if no response
+	// returned from etcd. Defaults to 30 seconds.
+	LeaderTimeout() time.Duration
+	SetLeaderTimeout(t time.Duration) ElectionOptions
+
+	// Duration after which a call to Resign() will timeout if no response
+	// returned from etcd. Defaults to 30 seconds.
+	ResignTimeout() time.Duration
+	SetResignTimeout(t time.Duration) ElectionOptions
+
+	// DefaultHostname returns the value that will be placed in the key of the
+	// leader of the election iff (1) no CampaionOptions option is passed to
+	// Campaign() and (2) the call to os.Hostname() fails. This value will very
+	// rarely be used since os.Hostname() should be reliable, and defaults to
+	// "default_hostname".
+	DefaultHostname() string
+	SetDefaultHostname(s string) ElectionOptions
+
+	// Hostname returns the hostname of the host if accessible, otherwise the
+	// value for DefaultHostname().
+	Hostname() string
+
+	String() string
+}
+
+// CampaignOptions provide the ability to override campaign defaults.
+type CampaignOptions interface {
+	// LeaderValue allows the user to override the value a campaign announces
+	// (that is, the value an observer sees upon calling Leader()). This defaults to
+	// the hostname of the caller.
+	LeaderValue() string
+	SetLeaderValue(v string) CampaignOptions
+
+	String() string
+}
+
+// LeaderService provides access to etcd-backed leader elections.
+type LeaderService interface {
+	// Close closes the election service client entirely. No more campaigns can be
+	// started and any outstanding campaigns are closed.
+	Close() error
+
+	// Campaign proposes that the caller become the leader for a specified
+	// election, with its leadership being refreshed on an interval of ttl
+	// seconds. If ttl is 0 it will default to etcd's default of 60s. It returns
+	// a read-only channel of campaign status events that is closed when the
+	// user resigns leadership or the campaign is invalidated due to background
+	// session expiration (i.e. failing to refresh etcd leadership lease). The
+	// caller MUST consume this channel until it is closed or risk goroutine
+	// leaks. Users are encouraged to read the package docs of services/leader
+	// for advice on proper usage and common gotchas.
+	//
+	// The leader will announce its hostname to observers unless opts is non-nil
+	// and opts.LeaderValue() is non-empty.
+	//
+	// NOTE: Once a campaign for a given electionID has been started, if it is
+	// lost or resigned and restarted it will be bound to the same TTL as the
+	// original due to constraints of etcd TTL sessions. The same applies if a
+	// call to Campaign() is made after a call to Leader() with the same ttl.
+	Campaign(electionID string, ttl int, opts CampaignOptions) (<-chan campaign.Status, error)
```

Comment:
Do we have a use case to campaign with different ttls under same LeaderService client? Would it be cleaner to move this ttl to ElectionOptions? so we don't need the logic to check different ttl for the same electionID etc.